packages/puppeteer-core/src/bidi/core/Browser.ts

```diff
@@ -0,0 +1,150 @@
+  /** Emitted after the browser closes. */
+  closed: {
+    /** The reason for closing the browser. */
+    reason: string;
+  };
+  /** Emitted after the browser disconnects. */
+  disconnected: {
+    /** The reason for disconnecting the browser. */
+    reason: string;
+  };
+  /** Emitted when a shared worker is created. */
+  sharedworker: {
+    /** The realm of the shared worker. */
+    realm: SharedWorkerRealm;
+  };
+}> {
+  static async from(session: Session): Promise<Browser> {
+    const browser = new Browser(session);
+    await browser.#initialize();
+    return browser;
+  }
+
+  // keep-sorted start
+  #reason: string | undefined;
+  readonly #userContexts = new Map();
+  readonly session: Session;
+  // keep-sorted end
+
+  private constructor(session: Session) {
+    super();
+
+    // keep-sorted start
+    this.session = session;
+    // keep-sorted end
+
+    this.#userContexts.set('', UserContext.create(this, ''));
+  }
+
+  async #initialize() {
+    // ///////////////////////
+    // Connection listeners //
+    // ///////////////////////
+    const connection = this.#connection;
+    connection.on('script.realmCreated', info => {
+      if (info.type === 'shared-worker') {
+        // TODO: Create a SharedWorkerRealm.
+      }
+    });
+
+    // ///////////////////
+    // Parent listeners //
+    // ///////////////////
+    this.session.once('ended', ({reason}) => {
+      this.#reason = reason;
+      this.emit('disconnected', {reason});
+      this.removeAllListeners();
+    });
+
+    // //////////////////////////////
+    // Asynchronous initialization //
+    // //////////////////////////////
+    const {
+      result: {contexts},
+    } = await connection.send('browsingContext.getTree', {});
+
+    // Simulating events so contexts are created naturally.
+    for (const context of contexts) {
+      connection.emit('browsingContext.contextCreated', context);
+      if (context.children) {
+        contexts.push(...context.children);
+      }
+    }
+  }
+
+  get #connection() {
+    return this.session.connection;
+  }
+
+  get disposed(): boolean {
+    return this.#reason !== undefined;
+  }
+
+  get defaultUserContext(): UserContext {
+    // SAFETY: A UserContext is always created for the default context.
+    return this.#userContexts.get('')!;
+  }
+
+  get userContexts(): Iterable<UserContext> {
+    return this.#userContexts.values();
+  }
+
+  @throwIfDisposed((browser: Browser) => {
+    // SAFETY: By definition of `disposed`, `#reason` is defined.
+    return browser.#reason!;
+  })
+  async close(): Promise<void> {
+    await this.#connection.send('browser.close', {});
+    this.#reason = `Browser has already closed.`;
+    this.emit('closed', {reason: this.#reason});
+    this.removeAllListeners();
+  }
+
+  @throwIfDisposed((browser: Browser) => {
+    // SAFETY: By definition of `disposed`, `#reason` is defined.
+    return browser.#reason!;
+  })
+  async addPreloadScript(
```

Comment:
how did you decide that the preload script should be on the browser?

Reply:
`contexts` is allowed to be optional, so `addPreloadScript` can be triggered on the all contexts. The highest node that contains all contexts is the browser (IIUC).